hudi-spark-datasource/hudi-spark-common/src/main/scala/org/apache/hudi/cdc/HoodieCDCRDD.scala

```diff
@@ -0,0 +1,618 @@
+
+    /**
+     * The deserializer used to convert the CDC GenericRecord to Spark InternalRow.
+     */
+    private lazy val cdcRecordDeserializer: HoodieAvroDeserializer = {
+      sparkAdapter.createAvroDeserializer(cdcAvroSchema, cdcSparkSchema)
+    }
+
+    private lazy val projection: UnsafeProjection = generateUnsafeProjection(cdcSchema, requiredCdcSchema)
+
+    // Iterator on cdc file
+    private val cdcFileIter = split.commitToChanges.sortBy(_._1).iterator
+
+    // The instant that is currently being processed
+    private var currentInstant: HoodieInstant = _
+
+    // The change file that is currently being processed
+    private var currentChangeFile: HoodieCDCFileSplit = _
+
+    /**
+     * Two cases will use this to iterator the records:
+     * 1) extract the change data from the base file directly, including 'ADD_BASE_File' and 'REMOVE_BASE_File'.
+     * 2) when the type of cdc file is 'REPLACED_FILE_GROUP',
+     *    use this to trace the records that are converted from the '[[beforeImageRecords]]
+     */
+    private var recordIter: Iterator[InternalRow] = Iterator.empty
+
+    /**
+     * Only one case where it will be used is that extract the change data from log files for mor table.
+     * At the time, 'logRecordIter' will work with [[beforeImageRecords]] that keep all the records of the previous file slice.
+     */
+    private var logRecordIter: Iterator[(String, HoodieRecord[_ <: HoodieRecordPayload[_ <: HoodieRecordPayload[_ <: AnyRef]]])] = Iterator.empty
+
+    /**
+     * Only one case where it will be used is that extract the change data from cdc log files.
+     */
+    private var cdcLogRecordIterator: HoodieCDCLogRecordIterator = _
+
+    /**
+     * The next record need to be returned when call next().
+     */
+    protected var recordToLoad: InternalRow = _
+
+    /**
+     * The list of files to which 'beforeImageRecords' belong.
+     * Use it to determine if 'beforeImageRecords' contains all the required data that extract
+     * the change data from the current cdc file.
+     */
+    private val beforeImageFiles: mutable.ArrayBuffer[String] = mutable.ArrayBuffer.empty
+
+    /**
+     * Keep the before-image data. There cases will use this:
+     * 1) the cdc infer case is [[LOG_FILE]];
+     * 2) the cdc infer case is [[AS_IS]] and [[cdcSupplementalLoggingMode]] is 'op_key'.
+     */
+    private var beforeImageRecords: mutable.Map[String, GenericRecord] = mutable.Map.empty
+
+    /**
+     * Keep the after-image data. Only one case will use this:
+     * the cdc infer case is [[AS_IS]] and [[cdcSupplementalLoggingMode]] is 'op_key' or 'cdc_data_before'.
+     */
+    private var afterImageRecords: mutable.Map[String, InternalRow] = mutable.Map.empty
+
+    private def needLoadNextFile: Boolean = {
+      !recordIter.hasNext &&
+        !logRecordIter.hasNext &&
+        (cdcLogRecordIterator == null || !cdcLogRecordIterator.hasNext)
+    }
+
+    @tailrec final def hasNextInternal: Boolean = {
```

Comment:
I think we should split `CDCFileGroupIterator` into N iterators for every `HoodieCDCInferenceCase` to make it more manageable and easier to understand